### PR TITLE
Add Mode connection string option

### LINKED ISSUE #180: Remove hard dependency on Microsoft.Extensions.PlatformAbstractions for net45
Resolving relative paths can be done using `AppContext`.
DNX runtime detection, can try and dynamically load the assembly at runtime.

### LINKED ISSUE #182: Support named, in-memory databases
The native API supports this using connection strings like `file:cow-db?mode=memory&cache=shared`

### LINKED ISSUE #188: Relative path in connection string pointing to different locations
I'm using a two project setup as described here https://github.com/aspnet/Microsoft.Data.Sqlite/issues/55#issuecomment-132480553

config.json
````
    "Data": {
        "DefaultConnection": {
            "ConnectionString": "Filename=database.db"
        }
    },
````

Startup.cs
````
[...]
            var builder = new ConfigurationBuilder()
                 .SetBasePath(appEnv.ApplicationBasePath)
                 .AddJsonFile("config.json");
[...]

            services.AddEntityFramework()
                .AddSqlite()
                .AddDbContext<MyAppDbContext>(
                    options => options.UseSqlite(_configuration["Data:DefaultConnection:ConnectionString"]));

[...]
````
I have a custom command defined in the Program.cs allowing me to reset and reseed the database.

````
    private static void ResetDatabase()
    {
        var dbContext = _services.GetService<MyAppDbContext>();

        dbContext.Database.EnsureDeleted();

        dbContext.Database.EnsureCreated();
        
        Console.WriteLine("Database reset successful.");
    }
````

When I run my application from IIS Express the database gets created in :
`D:\MyApp\src\MyApp\wwwroot`

When I run the custom command it appears in:
`D:\MyApp\src\MyApp\`

### PR DESCRIPTION
**Pull request by bricelam**
This also...
* Removes the PlatformAbstractions dependency (resolves #180)
* Fixes #188

Resolves #182

### CODE REVIEW

File: src/Microsoft.Data.Sqlite/SqliteConnection.cs [hunk 1]
```diff
@@ -103,89 +104,103 @@ public override void Open()
...
+                    break;
+
+                case SqliteCacheMode.Private:
+                    flags |= SQLITE_OPEN_PRIVATECACHE;
+                    break;
+
+                default:
+                    Debug.Assert(
+                        ConnectionStringBuilder.Cache == SqliteCacheMode.Default,
+                        "ConnectionStringBuilder.Cache is not Unspecified.");
```
Comment (natemcmaster) [inline src/Microsoft.Data.Sqlite/SqliteConnection.cs]: `not Unspecified`? Not sure what this means.

File: src/Microsoft.Data.Sqlite/SqliteConnection.cs [hunk 2]
```diff
@@ -103,89 +104,103 @@ public override void Open()
...
+
+            OnOpened();
+        }
+
+        partial void OnOpened();
+
+#if NETCORE50
+        partial void OnOpened()
+        {
+            var temporaryFolder = CurrentApplicationData?.TemporaryFolder.Path;
```
Comment (natemcmaster) [inline src/Microsoft.Data.Sqlite/SqliteConnection.cs]: I believe we replace this with `Path.GetTempPath()`.

File: src/Microsoft.Data.Sqlite/Utilities/NativeLibraryLoader.cs [hunk 3]
```diff
@@ -81,7 +81,17 @@ public static bool IsLoaded(string dllName)
         }
 
         private static bool IsDNX()
-            => PlatformServices.Default?.Application.RuntimeFramework.Identifier == "DNX";
+        {
+            dynamic defaultPlatformServices =
```
Comment (natemcmaster) [inline src/Microsoft.Data.Sqlite/Utilities/NativeLibraryLoader.cs]: Clever hack. Wish there were a better way cuz this feels gross.

File: test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs [hunk 4]
```diff
@@ -175,6 +155,29 @@ public void Open_works()
         }
 
         [Fact]
+        public void Can_use_named_shareable_inMemory_databases()
```
Comment (natemcmaster) [inline test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: Would `Data Source=file:people?mode=memory&cache=shared` also work?

File: src/Microsoft.Data.Sqlite/SqliteConnection.cs (hunk 1 above)
Comment (bricelam) [inline src/Microsoft.Data.Sqlite/SqliteConnection.cs]: Oops, I renamed it to Default, but missed this assert.

File: test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs (hunk 4 above)
Comment (bricelam) [inline test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: Yep, you can always "drop down" to using URIs. This just shows the "supported" ADO.NETty pattern.

File: src/Microsoft.Data.Sqlite/Utilities/NativeLibraryLoader.cs (hunk 3 above)
Comment (bricelam) [inline src/Microsoft.Data.Sqlite/Utilities/NativeLibraryLoader.cs]: We may be able to just use `AppDomain.CurrentDomain.SetupInformation.TargetFrameworkName`. I'll investigate.

File: src/Microsoft.Data.Sqlite/SqliteConnection.cs (hunk 2 above)
Comment (bricelam) [inline src/Microsoft.Data.Sqlite/SqliteConnection.cs]: Will try. Much simpler.

Comment (bricelam) [inline src/Microsoft.Data.Sqlite/SqliteConnection.cs]: Hmm... They're slightly different directories. The SQLite documentation uses `ApplicationData.Current.TemporaryFolder.Path`. We'll stick with that for now.

File: src/Microsoft.Data.Sqlite/Utilities/NativeLibraryLoader.cs (hunk 3 above)
Comment (bricelam) [inline src/Microsoft.Data.Sqlite/Utilities/NativeLibraryLoader.cs]: No good. Too many nuances. We'll stick with this hack for now. This code path may go away with CLI.

**Comment by natemcmaster** [comment]
The only thing I see missing is tests for actually opening URI connection strings. Are they already in the tests somewhere? Otherwise, :shipit:

**Comment by bricelam** [comment]
lol, ok, ok, I'll add more tests... I just got lazy. :sheep:

**Comment by bricelam** [comment]
Added missing tests for ReadOnly, ReadWrite, and URIs.

**Comment by natemcmaster** [comment]
:ship: :it:

**Pull request merged**